### ISSUE TITLE
Make binlog support integration client side epoll

### ISSUE THREAD

**Issue by andreas-bok-sociomantic** [sociomantic/libdrizzle-redux#34]
The current implementation of **binlog** can't be used with a **NON_BLOCKING** connection. 
As a consequence clients can't integrate libdrizzle's binlog with eg. an `epoll` based architecture 

To resolve the problem the function [`drizzle_binlog_start`](https://github.com/sociomantic-tsunami/libdrizzle-redux/blob/master/libdrizzle/binlog.cc#L75) should be refactored so internal calls to drizzle functions are separated.

**Comment by leandro-lucarella-sociomantic** [comment]
Oh, damn! Is there any technical reason for that? (`binlog` connection not using `NONBLOCK`)

**Comment by andreas-bok-sociomantic** [comment]
```
> Oh, damn! Is there any technical reason for that? (binlog connection not using NONBLOCK)
```

The reason is not stated in the doc. Could be that a client would typically only use one process/connection

The error happens in [drizzle_binlog_start](github.com/sociomantic-tsunami/libdrizzle-redux/blob/master/libdrizzle/binlog.cc#L95) where the call to `drizzle_query` with **NON_BLOCKING** makes `ret` be `DRIZZLE_RETURN_IO_WAIT`. On the client side we expect `drizzle_binlog_start` to return `DRIZZLE_RETURN_EOF` after successfully reading all binary logs. 

``` cpp
drizzle_return_t drizzle_binlog_start(drizzle_binlog_st *binlog,...)
{
...
  // Hack in 5.6 to say that client support checksums
  result= drizzle_query(con, "SET @master_binlog_checksum='NONE'", 0, &ret);
  drizzle_result_free(result);
  if (ret != DRIZZLE_RETURN_OK)
  {
    return ret;
  }
...
}
```

**Comment by nemanja-boric-sociomantic** [comment]
The reason is that this custom method (in libdrizzle) was just not written with non-blocking IO in mind.

**Comment by nemanja-boric-sociomantic** [comment]
(As it contains from many composite calls not made in non-blocking fashion).

**Comment by andreas-bok-sociomantic** [comment]
FYI @leandro-lucarella-sociomantic, @mathias-baumann-sociomantic, @ben-palmer-sociomantic, @nemanja-boric-sociomantic  

I was looking into #40 in the weekend and found that there is a more fundamental problem with non-blocking connections in drizzle-redux: 

> https://bugs.launchpad.net/libdrizzle/+bug/1211567 
> 
> The drizzle_connect() call blocks even if libdrizzle has been put in non-blocking mode. It does a nonblocking connect() call, but then libdrizzle blocks internally on poll in connect_poll() before returning to the caller, even if drizzle_options_set_non_blocking() has been used.

The bug was addressed in the development branch https://code.launchpad.net/~wiml-omni/libdrizzle/nonblocking
However this was merged into the original libdrizzle-redux master after the release of `v5.1.4`, which we have forked this libdrizzle-redux from

This is in line with the observed behaviour when trying to change **binlog** to use a non-blocking connection. That is, the connection step is never completed let, thus causing nothing to happen when calling [`binlog_start`](https://github.com/sociomantic-tsunami/libdrizzle-redux/blob/dc5c511dc4799516f2b2e1689ab8a33f48b2634b/libdrizzle/binlog.cc#L75). 

So AFAICS we need to test if merging the branch https://code.launchpad.net/~wiml-omni/libdrizzle/nonblocking into this master can make binlog **support integration client side epoll**

**Comment by ben-palmer-sociomantic** [comment]
> So AFAICS we need to test if merging the branch https://code.launchpad.net/~wiml-omni/libdrizzle/nonblocking into this master can make binlog support integration client side epoll

That seems like a reasonable solution to me. We can add this code as a PR and test against it.

**Comment by andreas-bok-sociomantic** [comment]
The issue has been addressed by #53. However the patch comes at a price. It allows the client to use a non-blocking connection by registering the file descriptor with the async io mechanism of choice. But, but, but parsing of raw binlog data has to be done by the clien, i.e. the client needs to implement [drizzle_state_binlog_read()](https://github.com/sociomantic-tsunami/libdrizzle-redux/blob/master/libdrizzle/binlog.cc#L272). 
This is off course not optimal, since binlog capabilities are at the core of the library. 
However, changing binlog api to work seamlessly with non-blocking connections require quite a lot refactoring. 
We wish to address this in a future release.
